### ISSUE TITLE
"Socket is closed" error using httpclient

### ISSUE THREAD

**Issue by wolfgangmm** [eXist-db/exist#1709]
### What is the problem

On latest develop, when using httpclient to GET a resource which is returned as xs:base64Binary, a following util:binary-to-string will fail with a "Socket is closed" error.

### What did you expect
Sample query below works fine in eXist 3.6.1 and returns a JSON string.

### Describe how to reproduce or add a test
```xquery
xquery version "3.1";

import module namespace http="http://expath.org/ns/http-client";

let $url := "https://www.ssrq-sds-fds.ch/places-db-edit/views/loc-search.xq?query=sevelen"
let $request :=
    <http:request method="GET" href="{$url}"/>
let $response := http:send-request($request)
let $json := util:binary-to-string($response[2])
return
    $json
```

### Context information

- 3.7.0-SNAPSHOT 8685d2d4e 
- java version "1.8.0_162"

Relevant stacktrace:

```
Caused by: java.net.SocketException: Socket is closed
        at sun.security.ssl.SSLSocketImpl.checkEOF(SSLSocketImpl.java:1542) ~[?:1.8.0_162]
        at sun.security.ssl.AppInputStream.read(AppInputStream.java:95) ~[?:1.8.0_162]
        at org.apache.http.impl.io.SessionInputBufferImpl.streamRead(SessionInputBufferImpl.java:137) ~[httpcore-4.4.9.jar:4.4.9]
        at org.apache.http.impl.io.SessionInputBufferImpl.fillBuffer(SessionInputBufferImpl.java:153) ~[httpcore-4.4.9.jar:4.4.9]
        at org.apache.http.impl.io.SessionInputBufferImpl.read(SessionInputBufferImpl.java:206) ~[httpcore-4.4.9.jar:4.4.9]
        at org.apache.http.impl.io.ContentLengthInputStream.read(ContentLengthInputStream.java:176) ~[httpcore-4.4.9.jar:4.4.9]
        at org.apache.http.conn.EofSensorInputStream.read(EofSensorInputStream.java:135) ~[httpclient-4.5.4.jar:4.5.4]
        at org.exist.util.io.AbstractFilterInputStreamCache.read(AbstractFilterInputStreamCache.java:118) ~[exist.jar:3.7.0-SNAPSHOT]
        at org.exist.util.io.CachingFilterInputStream.read(CachingFilterInputStream.java:176) ~[exist.jar:3.7.0-SNAPSHOT]
        at org.exist.util.io.CachingFilterInputStream.read(CachingFilterInputStream.java:141) ~[exist.jar:3.7.0-SNAPSHOT]
        at org.exist.xquery.value.BinaryValueFromInputStream.streamBinaryTo(BinaryValueFromInputStream.java:67) ~[exist.jar:3.7.0-SNAPSHOT]
        at org.exist.xquery.functions.util.BinaryToString.binaryToString(BinaryToString.java:118) ~[exist.jar:3.7.0-SNAPSHOT]
```

**Comment by dizzzz** [comment]
Same java version for both situations? Hmmm interesting to figure out what has changed....

**Comment by joewiz** [comment]
I can confirm this bug affects me too. 

@dizzzz Correct, while testing with java 1.8.0_162, the same code that works in 3.6.1 fails in 4.1.0-SNAPSHOT (current develop) with a "Socket is closed" string.

**Comment by joewiz** [comment]
A counterpoint, showing that the problem above affects binary objects and not text or XML files:

```xquery
xquery version "3.1";

import module namespace file="http://expath.org/ns/file";

let $txt := hc:send-request(<hc:request href="https://tools.ietf.org/rfc/rfc3986.txt" method="GET"/>)[2] 
return
    file:write-text("/Users/joe/Desktop/rfc3986.txt", $txt, "UTF-8")
```

```xquery
xquery version "3.1";

import module namespace file="http://expath.org/ns/file";

let $xml := hc:send-request(<hc:request href="http://demo.exist-db.org/exist/apps/public-repo/feed.xml" method="GET"/>)[2] 
return
    file:write("/Users/joe/Desktop/feed.xml", $xml)
```

**Comment by dizzzz** [comment]
might `http` vs `https` be a factor?

**Comment by joewiz** [comment]
It appears not. I've posted the JSON to a temporary location accessible via HTTP, http://s3.amazonaws.com/static.history.state.gov/temp/loc-search.json, and receive a similar error (I'm running the query in eXide with 4.1.0-SNAPSHOT current develop):

```xquery
hc:send-request(
  <hc:request 
    href="http://s3.amazonaws.com/static.history.state.gov/temp/loc-search.json" method="GET"/>
)[2]
```

eXide's session.xql returns this error when trying to execute the query:

> exerr:ERROR Unable to encode string value: The underlying InputStream has been closed

The corresponding error in exist.log:

```text
2018-03-01 14:47:56,078 [qtp1660359935-143] ERROR (XQueryContext.java [cleanup]:3613) - Unable to close binary value: Premature end of Content-Length delimited message body (expected: 263; received: 0 
org.apache.http.ConnectionClosedException: Premature end of Content-Length delimited message body (expected: 263; received: 0
	at org.apache.http.impl.io.ContentLengthInputStream.read(ContentLengthInputStream.java:178) ~[httpcore-4.4.9.jar:4.4.9]
	at org.apache.http.impl.io.ContentLengthInputStream.read(ContentLengthInputStream.java:198) ~[httpcore-4.4.9.jar:4.4.9]
	at org.apache.http.impl.io.ContentLengthInputStream.close(ContentLengthInputStream.java:101) ~[httpcore-4.4.9.jar:4.4.9]
	at org.apache.http.impl.execchain.ResponseEntityProxy.streamClosed(ResponseEntityProxy.java:143) ~[httpclient-4.5.4.jar:4.5.4]
	at org.apache.http.conn.EofSensorInputStream.checkClose(EofSensorInputStream.java:228) ~[httpclient-4.5.4.jar:4.5.4]
	at org.apache.http.conn.EofSensorInputStream.close(EofSensorInputStream.java:172) ~[httpclient-4.5.4.jar:4.5.4]
	at org.exist.util.io.AbstractFilterInputStreamCache.close(AbstractFilterInputStreamCache.java:81) ~[exist.jar:4.1.0-SNAPSHOT]
	at org.exist.util.io.CachingFilterInputStream.close(CachingFilterInputStream.java:200) ~[exist.jar:4.1.0-SNAPSHOT]
	at org.exist.xquery.value.BinaryValueFromInputStream.close(BinaryValueFromInputStream.java:97) ~[exist.jar:4.1.0-SNAPSHOT]
	at org.exist.xquery.XQueryContext$BinaryValueCleanupTask.cleanup(XQueryContext.java:3606) [exist.jar:4.1.0-SNAPSHOT]
	at org.exist.xquery.XQueryContext.runCleanupTasks(XQueryContext.java:3729) [exist.jar:4.1.0-SNAPSHOT]
	at org.exist.xquery.value.BinaryValueManager.runCleanupTasks(BinaryValueManager.java:14) [exist.jar:4.1.0-SNAPSHOT]
	at org.exist.http.servlets.XQueryServlet.process(XQueryServlet.java:473) [exist-optional.jar:4.1.0-SNAPSHOT]
	at org.exist.http.servlets.XQueryServlet.doPost(XQueryServlet.java:191) [exist-optional.jar:4.1.0-SNAPSHOT]
...
```

**Comment by dizzzz** [comment]
where can i find  the module "http://expath.org/ns/file"? #daretoask

**Comment by dizzzz** [comment]
@joewiz the query

```xquery
hc:send-request(
  <hc:request 
    href="http://s3.amazonaws.com/static.history.state.gov/temp/loc-search.json" method="GET"/>
)
```

yields for me into the following result:

```xml
<hc:response xmlns:hc="http://expath.org/ns/http-client" status="200" message="OK" spent-millis="142">
    <hc:header name="x-amz-id-2" value="PQbR+TQxnAlPNaNRAlIaCtINjjC/zbayW64/KcV3aIZfIkO4S5ZV4DfrJpP9us5TBNRUeeE112M="/>
    <hc:header name="x-amz-request-id" value="CB600D96C57E457B"/>
    <hc:header name="date" value="Thu, 01 Mar 2018 20:02:25 GMT"/>
    <hc:header name="last-modified" value="Thu, 01 Mar 2018 19:43:29 GMT"/>
    <hc:header name="etag" value="&#34;7032758437662a39a9db77b7c23a5005&#34;"/>
    <hc:header name="accept-ranges" value="bytes"/>
    <hc:header name="content-type" value="application/json"/>
    <hc:header name="content-length" value="263"/>
    <hc:header name="server" value="AmazonS3"/>
    <hc:body media-type="application/json"/>
</hc:response>
```

do I miss something? what do I wrong?

**Comment by joewiz** [comment]
You removed `[2]` from the end of the query, so what you're seeing is the 1st item returned by the `send-request` function. That's fine, but notice that the result indicator says "Query returned 2 results". Where's the second one? It's not being shown. eXide doesn't indicate there was any error loading the 2nd result. To see the actual error with your version of the code, open up the Developer Tools > Network pane, and you will see that the request to /results/2 fails, with this as the actual response body to /results/2:

```xml
<?xml version="1.0" ?>
<exception>
    <path>/db/apps/eXide/modules/session.xql</path>
    <message>exerr:ERROR Unable to encode string value: The underlying InputStream has been closed
        [at line 37, column 32, source: /db/apps/eXide/modules/session.xql] In function:
        local:retrieve(xs:integer) [104:24:/db/apps/eXide/modules/session.xql]</message>
</exception>
```

**Comment by dizzzz** [comment]
in the client I could fetch a more detailed stack trace

```
org.xmldb.api.base.XMLDBException: exerr:ERROR Unable to encode string value: The underlying InputStream has been closed
	at org.exist.xmldb.LocalXMLResource.getContent(LocalXMLResource.java:132)
	at org.exist.client.QueryDialog$QueryThread.run(QueryDialog.java:595)
Caused by: org.exist.xquery.XPathException: exerr:ERROR Unable to encode string value: The underlying InputStream has been closed
	at org.exist.xquery.value.BinaryValue.getStringValue(BinaryValue.java:222)
	at org.exist.xmldb.LocalXMLResource.getContent(LocalXMLResource.java:129)
	... 1 more
Caused by: java.io.IOException: The underlying InputStream has been closed
	at org.exist.util.io.CachingFilterInputStream.read(CachingFilterInputStream.java:138)
	at org.exist.util.io.CachingFilterInputStream.read(CachingFilterInputStream.java:131)
	at org.exist.xquery.value.BinaryValueFromInputStream.streamBinaryTo(BinaryValueFromInputStream.java:67)
	at org.exist.xquery.value.BinaryValue.streamTo(BinaryValue.java:251)
	at org.exist.xquery.value.BinaryValue.getStringValue(BinaryValue.java:220)
	... 2 more
Caused by: org.exist.xquery.XPathException: exerr:ERROR Unable to encode string value: The underlying InputStream has been closed
	at org.exist.xquery.value.BinaryValue.getStringValue(BinaryValue.java:222)
	at org.exist.xmldb.LocalXMLResource.getContent(LocalXMLResource.java:129)
	at org.exist.client.QueryDialog$QueryThread.run(QueryDialog.java:595)
Caused by: java.io.IOException: The underlying InputStream has been closed
	at org.exist.util.io.CachingFilterInputStream.read(CachingFilterInputStream.java:138)
	at org.exist.util.io.CachingFilterInputStream.read(CachingFilterInputStream.java:131)
	at org.exist.xquery.value.BinaryValueFromInputStream.streamBinaryTo(BinaryValueFromInputStream.java:67)
	at org.exist.xquery.value.BinaryValue.streamTo(BinaryValue.java:251)
	at org.exist.xquery.value.BinaryValue.getStringValue(BinaryValue.java:220)
	... 2 more
```

**Comment by dizzzz** [comment]
Interesting: if I store the query in the database, this is the (correct!?) response; no stack trace

```
<hc:response xmlns:hc="http://expath.org/ns/http-client" status="200" message="OK" spent-millis="295">
    <hc:header name="x-amz-id-2" value="Cthdf+QSEm7uEi9Yl0kurkBBuf5gjLg+GLQ0FRcaMU21An1rz92qwC+lnhQvWHNWxH699/RYhco="/>
    <hc:header name="x-amz-request-id" value="9785FD7DA989EFB5"/>
    <hc:header name="date" value="Thu, 01 Mar 2018 20:36:07 GMT"/>
    <hc:header name="last-modified" value="Thu, 01 Mar 2018 19:43:29 GMT"/>
    <hc:header name="etag" value="&#34;7032758437662a39a9db77b7c23a5005&#34;"/>
    <hc:header name="accept-ranges" value="bytes"/>
    <hc:header name="content-type" value="application/json"/>
    <hc:header name="content-length" value="263"/>
    <hc:header name="server" value="AmazonS3"/>
    <hc:body media-type="application/json"/>
</hc:response>eyAicmVzdWx0cyIgOiBbeyAiaWQiIDogImxvYzAwNjA0OSIsICJuYW1lIiA6ICJTZXZlbGVuIEhvZiIgfSwgeyAiaWQiIDogImxvYzAwNzc0NSIsICJuYW1lIiA6ICJTZXZlbGVuIEtpcmNoZSIgfSwgeyAiaWQiIDogImxvYzAwNjE3MSIsICJuYW1lIiA6ICJTZXZlbGVuIERvcmYiIH0sIHsgImlkIiA6ICJsb2MwMDA3MDgiLCAibmFtZSIgOiAiU2V2ZWxlbiIgfSwgeyAiaWQiIDogImxvYzAwNjAzOSIsICJuYW1lIiA6ICJTZXZlbGVuIEtpcmNoc3BpZWwiIH1dIH0=
```

**Comment by joewiz** [comment]
@dizzzz Sorry, I missed your earlier question about where to get the EXPath File Module. I built it from https://github.com/eXist-db/expath-file-module. (I was using it to rule out eXist's native file module as the source of the problem. I really like the EXPath File Module and would love to have this available more easily. The more cross-platform our code can be the better. Thus I've opened https://github.com/eXist-db/expath-file-module/issues/6 to discuss this.
